core/vm/contracts.go

```diff
@@ -1353,3 +1358,92 @@ func (c *p256Verify) Run(input []byte) ([]byte, error) {
 		return nil, nil
 	}
 }
+
+// gpgEd25519Verify implements native verification for ed25519 signatures produced via gpg
+type gpgEd25519Verify struct{}
+
+var (
+	errMessageTooShort    = errors.New("message too short")
+	errPubKeyTooShort     = errors.New("public key too short")
+	errSignatureTooShort  = errors.New("signature too short")
+	errInvalidPublicKey   = errors.New("invalid public key format")
+	errInvalidSignature   = errors.New("invalid signature format")
+	errVerificationFailed = errors.New("signature verification failed")
+)
+
+// RequiredGas returns the gas required to execute the pre-compiled contract
+func (c *gpgEd25519Verify) RequiredGas(input []byte) uint64 {
+	// You can adjust this value based on your needs
+	return params.GpgEd25519VerifyGas
+}
+
+// Run performs ed25519 signature verification
+// Q_TODO: Enforce max message size?
+func (c *gpgEd25519Verify) Run(input []byte) ([]byte, error) {
+	// Input should be: message_len (32 bytes) || message || pubkey_len (32 bytes) || pubkey || sig_len (32 bytes) || signature
+	if len(input) < 96 { // minimum length for the three length fields
+		return nil, errMessageTooShort
+	}
+
+	// Extract message length and message
+	msgLen := new(big.Int).SetBytes(input[:32]).Uint64()
+	if len(input) < 32+int(msgLen) {
+		return nil, errMessageTooShort
+	}
+	message := input[32 : 32+msgLen]
+
+	// Extract public key length and public key
+	offset := 32 + msgLen
+	pubKeyLen := new(big.Int).SetBytes(input[offset : offset+32]).Uint64()
+	if len(input) < int(offset+32+pubKeyLen) {
+		return nil, errPubKeyTooShort
+	}
+	pubKey := input[offset+32 : offset+32+pubKeyLen]
+
+	// Extract signature length and signature
+	offset = offset + 32 + pubKeyLen
+	sigLen := new(big.Int).SetBytes(input[offset : offset+32]).Uint64()
+	if len(input) < int(offset+32+sigLen) {
+		return nil, errSignatureTooShort
+	}
+	signature := input[offset+32 : offset+32+sigLen]
+
+	// Convert raw bytes to armored format
+	armoredPubKey := string(pubKey)
+	armoredSig := string(signature)
+
+	// Create public key object
+	pubKeyObj, err := pgpcrypto.NewKeyFromArmored(armoredPubKey)
+	if err != nil {
+		return nil, errInvalidPublicKey
+	}
+
+	// Create public keyring
+	pubKeyRing, err := pgpcrypto.NewKeyRing(pubKeyObj)
+	if err != nil {
+		return nil, errInvalidPublicKey
+	}
+
+	// Parse the armored signature
+	signatureObj, err := pgpcrypto.NewPGPSignatureFromArmored(armoredSig)
+	if err != nil {
+		return nil, errInvalidSignature
+	}
+
+	// Create message object
+	messageHex := hex.EncodeToString(message)
+	messageBytes, err := hex.DecodeString(messageHex)
+	if err != nil {
+		return nil, err
+	}
+	messageObj := pgpcrypto.NewPlainMessage(messageBytes)
+
+	// Verify signature
+	err = pubKeyRing.VerifyDetached(messageObj, signatureObj, pgpcrypto.GetUnixTime())
+	if err != nil {
+		return nil, errVerificationFailed
```

Comment:
Signature verification failure should just return `bytes32(0)` instead of an error.